// @azure/msal-node/dist/index.js
/*! @azure/msal-node v2.0.0-beta.1 2023-07-06 */
'use strict';
'use strict';

var internals = require('./internals.js');
var PublicClientApplication = require('./client/PublicClientApplication.js');
var ConfidentialClientApplication = require('./client/ConfidentialClientApplication.js');
var ClientApplication = require('./client/ClientApplication.js');
var ClientCredentialClient = require('./client/ClientCredentialClient.js');
var DeviceCodeClient = require('./client/DeviceCodeClient.js');
var OnBehalfOfClient = require('./client/OnBehalfOfClient.js');
var UsernamePasswordClient = require('./client/UsernamePasswordClient.js');
var Configuration = require('./config/Configuration.js');
var ClientAssertion = require('./client/ClientAssertion.js');
var TokenCache = require('./cache/TokenCache.js');
var NodeStorage = require('./cache/NodeStorage.js');
var DistributedCachePlugin = require('./cache/distributed/DistributedCachePlugin.js');
var CryptoProvider = require('./crypto/CryptoProvider.js');
var msalCommon = require('@azure/msal-common');
var packageMetadata = require('./packageMetadata.js');



exports.internals = internals;
exports.PublicClientApplication = PublicClientApplication.PublicClientApplication;
exports.ConfidentialClientApplication = ConfidentialClientApplication.ConfidentialClientApplication;
exports.ClientApplication = ClientApplication.ClientApplication;
exports.ClientCredentialClient = ClientCredentialClient.ClientCredentialClient;
exports.DeviceCodeClient = DeviceCodeClient.DeviceCodeClient;
exports.OnBehalfOfClient = OnBehalfOfClient.OnBehalfOfClient;
exports.UsernamePasswordClient = UsernamePasswordClient.UsernamePasswordClient;
exports.buildAppConfiguration = Configuration.buildAppConfiguration;
exports.ClientAssertion = ClientAssertion.ClientAssertion;
exports.TokenCache = TokenCache.TokenCache;
exports.NodeStorage = NodeStorage.NodeStorage;
exports.DistributedCachePlugin = DistributedCachePlugin.DistributedCachePlugin;
exports.CryptoProvider = CryptoProvider.CryptoProvider;
Object.defineProperty(exports, 'AuthError', {
	enumerable: true,
	get: function () { return msalCommon.AuthError; }
});
Object.defineProperty(exports, 'AuthErrorMessage', {
	enumerable: true,
	get: function () { return msalCommon.AuthErrorMessage; }
});
Object.defineProperty(exports, 'AzureCloudInstance', {
	enumerable: true,
	get: function () { return msalCommon.AzureCloudInstance; }
});
Object.defineProperty(exports, 'ClientAuthError', {
	enumerable: true,
	get: function () { return msalCommon.ClientAuthError; }
});
Object.defineProperty(exports, 'ClientAuthErrorMessage', {
	enumerable: true,
	get: function () { return msalCommon.ClientAuthErrorMessage; }
});
Object.defineProperty(exports, 'ClientConfigurationError', {
	enumerable: true,
	get: function () { return msalCommon.ClientConfigurationError; }
});
Object.defineProperty(exports, 'ClientConfigurationErrorMessage', {
	enumerable: true,
	get: function () { return msalCommon.ClientConfigurationErrorMessage; }
});
Object.defineProperty(exports, 'InteractionRequiredAuthError', {
	enumerable: true,
	get: function () { return msalCommon.InteractionRequiredAuthError; }
});
Object.defineProperty(exports, 'InteractionRequiredAuthErrorMessage', {
	enumerable: true,
	get: function () { return msalCommon.InteractionRequiredAuthErrorMessage; }
});
Object.defineProperty(exports, 'LogLevel', {
	enumerable: true,
	get: function () { return msalCommon.LogLevel; }
});
Object.defineProperty(exports, 'Logger', {
	enumerable: true,
	get: function () { return msalCommon.Logger; }
});
Object.defineProperty(exports, 'PromptValue', {
	enumerable: true,
	get: function () { return msalCommon.PromptValue; }
});
Object.defineProperty(exports, 'ProtocolMode', {
	enumerable: true,
	get: function () { return msalCommon.ProtocolMode; }
});
Object.defineProperty(exports, 'ResponseMode', {
	enumerable: true,
	get: function () { return msalCommon.ResponseMode; }
});
Object.defineProperty(exports, 'ServerError', {
	enumerable: true,
	get: function () { return msalCommon.ServerError; }
});
Object.defineProperty(exports, 'TokenCacheContext', {
	enumerable: true,
	get: function () { return msalCommon.TokenCacheContext; }
});
exports.version = packageMetadata.version;
